bft_interp/src/lib.rs

```diff
@@ -1,39 +1,510 @@
+        let vm: VirtualMachine<u8> = VirtualMachine::new(&test_program, None, true);
+
+        assert_eq!(vm.cells.len(), 30_000);
+    }
+
+    // Does moving the head left with space on an extensible tape work?
+    #[test]
+    fn test_move_head_left_extensible_good() {
+        let test_program = test_program();
+        let mut vm: VirtualMachine<u8> = VirtualMachine::new(&test_program, None, true);
+        vm.head = 5;
+
+        let result = vm.move_head_left();
+
+        assert!(result.is_ok());
+        assert_eq!(vm.head, 4);
+    }
+
+    // Does moving the head left with space on an fixed tape work?
+    #[test]
+    fn test_move_head_left_fixed_good() {
+        let test_program = test_program();
+        let mut vm: VirtualMachine<u8> = VirtualMachine::new(&test_program, None, false);
+        vm.head = 5;
+
+        let result = vm.move_head_left();
+
+        assert!(result.is_ok());
+        assert_eq!(vm.head, 4);
+    }
+
+    // Does moving the head left at the start of an extensible tape error correctly?
+    #[test]
+    fn test_move_head_left_extensible_bad() {
+        let test_program = test_program();
+        let bad_instruction = test_program.instructions()[0].clone();
+
+        let mut vm: VirtualMachine<u8> = VirtualMachine::new(&test_program, None, true);
+
+        let result = vm.move_head_left();
+        let expected_error: Result<(), VMError> = Err(VMError::HeadUnderrun(bad_instruction));
+
+        assert!(result.is_err());
+        assert_eq!(result, expected_error);
+    }
+
+    // Does moving the head left at the start of an fixed tape error correctly?
+    #[test]
+    fn test_move_head_left_fixed_bad() {
+        let test_program = test_program();
+        let bad_instruction = test_program.instructions()[0].clone();
+
+        let mut vm: VirtualMachine<u8> = VirtualMachine::new(&test_program, None, false);
+
+        let result = vm.move_head_left();
+        let expected_error: Result<(), VMError> = Err(VMError::HeadUnderrun(bad_instruction));
+
+        assert!(result.is_err());
+        assert_eq!(result, expected_error);
+    }
+
+    // Does moving the head right on an extensible tape work when the head has space to move?
+    #[test]
+    fn test_move_head_right_extensible_good() {
+        let test_program = test_program();
+        let tape_len = Some(NonZeroUsize::new(1000).unwrap());
+        let mut vm: VirtualMachine<u8> = VirtualMachine::new(&test_program, tape_len, true);
+
+        let result = vm.move_head_right();
+
+        assert!(result.is_ok());
+        assert_eq!(vm.head, 1);
+    }
+
+    // Does moving the head right on an fixed tape work?
+    #[test]
+    fn test_move_head_right_fixed_good() {
+        let test_program = test_program();
+        let mut vm: VirtualMachine<u8> = VirtualMachine::new(&test_program, None, false);
+
+        let result = vm.move_head_right();
+
+        assert!(result.is_ok());
+        assert_eq!(vm.head, 1);
+    }
```

Comment:
If you're going to duplicate these functions, then at least check that the move-right didn't unnecessarily extend the tape.